src/Features/Core/Portable/EmbeddedLanguages/RegularExpressions/LanguageServices/RegexLanguageDetector.cs

```diff
@@ -0,0 +1,230 @@
+        {
+            var result = syntaxFacts.IsCaseSensitive
+                ? new HashSet<string>()
+                : new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+
+            if (regexType != null)
+            {
+                result.AddRange(
+                    from method in regexType.GetMembers().OfType<IMethodSymbol>()
+                    where method.DeclaredAccessibility == Accessibility.Public
+                    where method.IsStatic
+                    where method.Parameters.Any(p => p.Name == _patternName)
+                    select method.Name);
+            }
+
+            return result;
+        }
+
+        protected override bool IsEmbeddedLanguageString(
+            SyntaxToken token,
+            SyntaxNode argumentNode,
+            SemanticModel semanticModel,
+            CancellationToken cancellationToken,
+            out RegexOptions options)
+        {
+            if (_regexType == null)
+            {
+                options = default;
+                return false;
+            }
+
+            var syntaxFacts = Info.SyntaxFacts;
+            var argumentList = argumentNode.GetRequiredParent();
+            var invocationOrCreation = argumentList.Parent;
+            if (syntaxFacts.IsInvocationExpression(invocationOrCreation))
+            {
+                var invokedExpression = syntaxFacts.GetExpressionOfInvocationExpression(invocationOrCreation);
+                var name = GetNameOfInvokedExpression(invokedExpression);
+                if (name != null && _methodNamesOfInterest.Contains(name))
+                {
+                    // Is a string argument to a method that looks like it could be a Regex method.  
+                    // Need to do deeper analysis.
+
+                    // Note we do not use GetAllSymbols here because we don't want to incur the
+                    // allocation.
+                    var symbolInfo = semanticModel.GetSymbolInfo(invocationOrCreation, cancellationToken);
+                    var method = symbolInfo.Symbol;
+                    if (TryAnalyzeInvocation(_regexType, argumentNode, semanticModel, method, cancellationToken, out options))
+                        return true;
+
+                    foreach (var candidate in symbolInfo.CandidateSymbols)
+                    {
+                        if (TryAnalyzeInvocation(_regexType, argumentNode, semanticModel, candidate, cancellationToken, out options))
+                            return true;
+                    }
+                }
+            }
+            else if (syntaxFacts.IsObjectCreationExpression(invocationOrCreation))
+            {
+                var typeNode = syntaxFacts.GetTypeOfObjectCreationExpression(invocationOrCreation);
+                var name = GetNameOfType(typeNode, syntaxFacts);
+                if (name != null)
+                {
+                    if (syntaxFacts.StringComparer.Compare(nameof(Regex), name) == 0)
```

Comment:
We're just ignoring aliases here?

Reply:
We've had discussions around source generators too where a "is the type here being created" helper that does this might be helpful, not sure if we should extract this pattern. At least that way the ignoring of aliases can be in one place so we can fix it everywhere later?

Reply:
Yes we do.  Note: this is going to go semantic in a followup PR this week to support StringSyntaxAttribute.  So this is just about detecting this when older frameworks.